core/coins/format.go

```diff
@@ -0,0 +1,92 @@
+package coins
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	bankv1beta1 "cosmossdk.io/api/cosmos/bank/v1beta1"
+	basev1beta1 "cosmossdk.io/api/cosmos/base/v1beta1"
+	"cosmossdk.io/math"
+)
+
+// formatCoin formats a sdk.Coin into a value-rendered string, using the
+// given metadata about the denom. It returns the formatted coin string, the
+// display denom, and an optional error.
+func formatCoin(coin *basev1beta1.Coin, metadata *bankv1beta1.Metadata) (string, error) {
+	coinDenom := coin.Denom
+
+	// Return early if no display denom or display denom is the current coin denom.
+	if metadata == nil || metadata.Display == "" || coinDenom == metadata.Display {
+		vr, err := math.FormatDec(coin.Amount)
+		return vr + " " + coin.Denom, err
+	}
+
+	dispDenom := metadata.Display
+
+	// Find exponents of both denoms.
+	var coinExp, dispExp uint32
+	foundCoinExp, foundDispExp := false, false
+	for _, unit := range metadata.DenomUnits {
+		if coinDenom == unit.Denom {
+			coinExp = unit.Exponent
+			foundCoinExp = true
+		}
+		if dispDenom == unit.Denom {
```

Comment:
Why are we comparing against `unit.Denom` here? shouldn't it be `unit.Display`?

Reply:
There's no `unit.Display`, the metadata object looks like
```json
{
  "metadata": {
    "description": "The native staking token of the Cosmos Hub.",
    "denom_units": [
      {
        "denom": "uatom",
        "exponent": 0,
        "aliases": [
          "microatom"
        ]
      },
      {
        "denom": "matom",
        "exponent": 3,
        "aliases": [
          "milliatom"
        ]
      },
      {
        "denom": "atom",
        "exponent": 6,
        "aliases": [
        ]
      }
    ],
    "base": "uatom",
    "display": "atom",
    "name": "",
    "symbol": ""
  }
}
```

This piece of code finds the denom unit object (which contains the exponent) from inside the `denom_units` array